dvc/scm/git/__init__.py

```diff
@@ -229,7 +229,8 @@ def is_tracked(self, path):
         # it is equivalent to `bool(self.repo.git.ls_files(path))` by
         # functionality, but ls_files fails on unicode filenames
         path = relpath(path, self.root_dir)
-        return path in [i[0] for i in self.repo.index.entries]
+        # There are 4 stages, see BaseIndexEntry.stage
+        return any((path, i) in self.repo.index.entries for i in (0, 1, 2, 3))
```

Comment:
There might be a new index version released at some point and this might break in an non-obvious way.

Reply:
Do you you have any benchmark-ish results on how much problems this is causing us right now?

Reply:
These are not index versions. From `BaseIndexEntry.stage` docstring:
```
Stage of the entry, either:

    * 0 = default stage
    * 1 = stage before a merge or common ancestor entry in case of a 3 way merge
    * 2 = stage of entries from the 'left' side of the merge
    * 3 = stage of entries from the right side of the merge

:note: For more information, see http://www.kernel.org/pub/software/scm/git/docs/git-read-tree.html
```

Reply:
@Suor How much real-world problems is it causing us? Does it justify this hack at all?

Reply:
Tbh, glimpsing at it I don't quite understand what is going on.

Reply:
This is the kind of issue, which will sit for a while and then bite somebody, who might never understand what is going on, but just drop dvc because it's slow.

Reply:
@Suor So you are saying you do have some benchmark to show where this is causing issues? If so, please show it and show the results before and after your change.

Reply:
Benched it doesn't help much, the reason is `.index` is not cached it's reread on each use.

Reply:
Listing 10k files in 10k git repo took 6:19 with old code and 5:16 with new one.

Reply:
@Suor When do we do that? When we have 10K stage outputs? We don't list files inside dirs, right?

Reply:
We list when we do `dvc add -r dir`.

Reply:
@Suor But we don't recomend doing that on thousands of files, and consider that a misuse. Hence why I'm questioning if we even need this change.